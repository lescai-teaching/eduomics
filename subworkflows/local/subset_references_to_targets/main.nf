include { SUBVAR                                              } from '../../../modules/local/subvar/main'
include { SUBSETCAPTURE                                       } from '../../../modules/local/subsetcapture/main'
include { SAMTOOLS_FAIDX as SAMTOOLS_FAIDX_SUBSET             } from '../../../modules/nf-core/samtools/faidx/main'
include { SAMTOOLS_FAIDX as SAMTOOLS_FAIDX_INDEX              } from '../../../modules/nf-core/samtools/faidx/main'
include { SAMTOOLS_FAIDX as SAMTOOLS_FAIDX_SIZES              } from '../../../modules/nf-core/samtools/faidx/main'
include { GATK4_SELECTVARIANTS as GATK4_SELECTVARIANTS_GNOMAD } from '../../../modules/nf-core/gatk4/selectvariants/main'
include { GATK4_SELECTVARIANTS as GATK4_SELECTVARIANTS_MILLS  } from '../../../modules/nf-core/gatk4/selectvariants/main'
include { GATK4_SELECTVARIANTS as GATK4_SELECTVARIANTS_1000G  } from '../../../modules/nf-core/gatk4/selectvariants/main'
include { GATK4_SELECTVARIANTS as GATK4_SELECTVARIANTS_DBSNP  } from '../../../modules/nf-core/gatk4/selectvariants/main'
include { GATK4_CREATESEQUENCEDICTIONARY                      } from '../../../modules/nf-core/gatk4/createsequencedictionary/main'



workflow SUBSET_REFERENCES_TO_TARGETS {

    take:
    ch_meta          // channel: [ val(meta)   ]
    ch_fasta         // channel: [ path(fasta) ]
    ch_fai           // channel: [ path(fai)   ]
    ch_get_sizes     // channel: [ get_sizes   ] // boolean
    ch_capture_bed   // channel: [ path(capture_bed) ]
    ch_gnomad        // channel: [ path(vcf), path(vcf_idx) ]
    ch_mills         // channel: [ path(vcf), path(vcf_idx) ]
    ch_1000g         // channel: [ path(vcf), path(vcf_idx) ]
    ch_dbsnp         // channel: [ path(vcf), path(vcf_idx) ]
    ch_clinvar_vcf   // channel: [ path(vcf) ]

    main:

    ch_versions = Channel.empty()

    ch_fasta_with_meta = ch_meta.combine(ch_fasta).map { meta, fasta -> [meta, fasta] }
    ch_fai_with_meta = ch_meta.combine(ch_fai).map { meta, fai -> [meta, fai] }

    SAMTOOLS_FAIDX_SUBSET ( ch_fasta_with_meta, [[],[]], ch_get_sizes )
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX_SUBSET.out.versions)

    SAMTOOLS_FAIDX_INDEX (SAMTOOLS_FAIDX_SUBSET.out.fa, [[],[]], ch_get_sizes )
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX_INDEX.out.versions)

    SAMTOOLS_FAIDX_SIZES ( SAMTOOLS_FAIDX_SUBSET.out.fa, SAMTOOLS_FAIDX_INDEX.out.fai, true )
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX_SIZES.out.versions)

    GATK4_CREATESEQUENCEDICTIONARY ( SAMTOOLS_FAIDX_SUBSET.out.fa )
    ch_versions = ch_versions.mix(GATK4_CREATESEQUENCEDICTIONARY.out.versions)

    //Subset capture regions by chrom
    SUBSETCAPTURE (
        ch_meta,
        SAMTOOLS_FAIDX_SIZES.out.sizes.map { meta, sizes -> sizes },
        ch_capture_bed
        )

    ch_versions = ch_versions.mix(SUBSETCAPTURE.out.versions)

    // Subset gatk bundle vcf
    GATK4_SELECTVARIANTS_GNOMAD ( SUBSETCAPTURE.out.target_bed.combine(ch_gnomad).map { meta, target_bed, vcf, vcf_idx ->
        [meta, vcf, vcf_idx, target_bed]
    })
    ch_versions = ch_versions.mix(GATK4_SELECTVARIANTS_GNOMAD.out.versions)

    GATK4_SELECTVARIANTS_MILLS ( SUBSETCAPTURE.out.target_bed.combine(ch_mills).map { meta, target_bed, vcf, vcf_idx ->
        [meta, vcf, vcf_idx, target_bed]
    })
    ch_versions = ch_versions.mix(GATK4_SELECTVARIANTS_MILLS.out.versions)

    GATK4_SELECTVARIANTS_1000G ( SUBSETCAPTURE.out.target_bed.combine(ch_1000g).map { meta, target_bed, vcf, vcf_idx ->
        [meta, vcf, vcf_idx, target_bed]
    })
    ch_versions = ch_versions.mix(GATK4_SELECTVARIANTS_1000G.out.versions)

    GATK4_SELECTVARIANTS_DBSNP (SUBSETCAPTURE.out.target_bed.combine(ch_dbsnp).map { meta, target_bed, vcf, vcf_idx ->
        [meta, vcf, vcf_idx, target_bed]
    })
    ch_versions = ch_versions.mix(GATK4_SELECTVARIANTS_DBSNP.out.versions)

    // Subset clinvar variants
    SUBVAR (
    ch_meta.combine(ch_clinvar_vcf).map { meta, vcf -> [meta, vcf] },
    SUBSETCAPTURE.out.target_bed.map { meta, target_bed -> target_bed }
    )

    ch_versions = ch_versions.mix(SUBVAR.out.versions)


    emit:
    target_fa         = SAMTOOLS_FAIDX_SUBSET.out.fa                                 // channel:  [ val(meta), [ fa, fasta ] ]
    target_fai        = SAMTOOLS_FAIDX_INDEX.out.fai                                 // channel: [ val(meta), [ fai ] ]
    target_sizes      = SAMTOOLS_FAIDX_SIZES.out.sizes                               // channel: [ val(meta), [ sizes ] ]
    capture_bed_gz    = SUBSETCAPTURE.out.capture_bed_gz.map { meta, bed -> bed }    // channel: [ [ capture_bed_gz ] ]
    capture_bed_index = SUBSETCAPTURE.out.capture_bed_index.map { meta, bed -> bed } // channel: [ [ capture_bed_index ] ]
    target_bed        = SUBSETCAPTURE.out.target_bed.map { meta, bed -> bed }        // channel: [ [ target_bed ] ]
    target_bed_pad50  = SUBSETCAPTURE.out.target_bed_pad50.map { meta, bed -> bed }  // channel: [ [ target_bed_pad50 ] ]
    target_bed_pad500 = SUBSETCAPTURE.out.target_bed_pad500.map { meta, bed -> bed } // channel: [ [ target_bed_pad500 ] ]
    target_gnomad_vcf = GATK4_SELECTVARIANTS_GNOMAD.out.vcf.map { meta, vcf -> vcf } // channel: [ [ vcf ] ]
    target_gnomad_tbi = GATK4_SELECTVARIANTS_GNOMAD.out.tbi.map { meta, tbi -> tbi } // channel: [ [ tbi ] ]
    target_mills_vcf  = GATK4_SELECTVARIANTS_MILLS.out.vcf.map { meta, vcf -> vcf }  // channel: [ [ vcf ] ]
    target_mills_tbi  = GATK4_SELECTVARIANTS_MILLS.out.tbi.map { meta, tbi -> tbi }  // channel: [ [ tbi ] ]
    target_1000g_vcf  = GATK4_SELECTVARIANTS_1000G.out.vcf.map { meta, vcf -> vcf }  // channel: [ [ vcf ] ]
    target_1000g_tbi  = GATK4_SELECTVARIANTS_1000G.out.tbi.map { meta, tbi -> tbi }  // channel: [ [ tbi ] ]
    target_dbsnp_vcf  = GATK4_SELECTVARIANTS_DBSNP.out.vcf.map { meta, vcf -> vcf }  // channel: [ [ vcf ] ]
    target_dbsnp_tbi  = GATK4_SELECTVARIANTS_DBSNP.out.tbi.map { meta, tbi -> tbi }  // channel: [ [ tbi ] ]
    clinvar_benign_vcf      = SUBVAR.out.benign_vcf                                  // channel: [ val(meta), [ benign_vcf ] ]
    clinvar_pathogenic_vcf  = SUBVAR.out.pathogenic_vcf                              // channel: [ val(meta), [ pathogenic_vcf ] ]
    clinvar_selected_vcf    = SUBVAR.out.selected_vcf                                // channel: [ val(meta), [ selected_vcf ] ]
    target_dict = GATK4_CREATESEQUENCEDICTIONARY.out.dict
    versions = ch_versions                                                           // channel: [ versions.yml ]
}



