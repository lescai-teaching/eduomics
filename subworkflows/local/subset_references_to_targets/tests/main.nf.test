// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
// nf-core subworkflows test subset_references_to_targets
nextflow_workflow {

    name "Test Subworkflow SUBSET_REFERENCES_TO_TARGETS"
    script "../main.nf"
    config "subworkflows/local/subset_references_to_targets/tests/test.config"
    workflow "SUBSET_REFERENCES_TO_TARGETS"

    tag "subworkflows"
    tag "subworkflows/subset_references_to_targets"
    tag "subworkflow/subvar"
    tag "samtools"
    tag "samtools/faidx"
    tag "gatk4/"
    tag "gatk4/selectvariants"


    test("human - fasta/BED/VCF (dbsnp) - subset by chr") {

        when {
            workflow {
                """
                // ch_fasta        [mandatory] [ val(meta), [ fasta ] ]
                input[0] = Channel.value([
                    [ id:'test', chromosome:'chr22' ],
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/chr22_q11-21_GRCh38.fasta', checkIfExists: true)
                ])
                // ch_fai          [optional]  [ val(meta), [fai] ]
                input[1] = Channel.value([
                    [ id:'test', chromosome:'chr22' ],
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/chr22_q11-21_GRCh38.fasta.fai', checkIfExists: true)
                ])
                // ch_get_sizes
                input[2] = Channel.value(true)

                // ch_capture_bed - twist capture for exome
                input[3] = Channel.value(file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/Twist_Exome_Target2_chr22q11-21_GRCh38.bed', checkIfExists: true))

                // ch_chromosome
                input[4] = Channel.value([ [ id:'test' ], 'chr22' ])

                // ch_dbsnp
                input[5] = Channel.value([
                    [ id:'test' ],
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/dbsnp_146_chr22_q11-21_fixed.vcf.gz', checkIfExists: true),
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/dbsnp_146_chr22_q11-21_fixed.vcf.gz', checkIfExists: true),
                    file('https://raw.githubusercontent.com/lescai-teaching/eduomics_testdata/refs/heads/main/dna/Twist_Exome_Target2_chr22q11-21_GRCh38.bed', checkIfExists: true)
                ])
                // ch_gnomad
                input[5] = Channel.value([
                    [ id:'gnomad' ],
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/af-only-gnomad_chr22_q11-21_GRCh38.vcf.gz', checkIfExists: true),
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/af-only-gnomad_chr22_q11-21_GRCh38.vcf.gz.tbi', checkIfExists: true)
                ])
                // ch_mills
                input[6] = Channel.value([
                    [ id:'mills' ],
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/Mills_and_1000G_gold_standard.indels.hg38_chr22q11-21_fixed.vcf.gz', checkIfExists: true),
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/Mills_and_1000G_gold_standard.indels.hg38_chr22q11-21_fixed.vcf.gz.tbi', checkIfExists: true)
                ])
                // ch_1000g
                input[7] = Channel.value([
                    [ id:'1000g' ],
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/1000G_omni2_5_chr22_q11-21_GRCh38.vcf.gz', checkIfExists: true),
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/gatk_bundle/1000G_omni2_5_chr22_q11-21_GRCh38.vcf.gz.tbi', checkIfExists: true)
                ])
                // ch_clinvar_vcf
                input[6] = Channel.value([
                    [ id:'clinvar' ],
                    file('https://github.com/lescai-teaching/eduomics_testdata/blob/main/dna/clinvar_data/clinvar_20250317_chr22q11-21.vcf.gz', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.target_fa },
                { assert workflow.out.target_fai },
                { assert workflow.out.target_sizes },
                { assert workflow.out.capture_bed_gz },
                { assert workflow.out.capture_bed_index },
                { assert workflow.out.target_bed },
                { assert workflow.out.target_bed_pad50 },
                { assert workflow.out.target_bed_pad500 },
                { assert workflow.out.target_gnomad_vcf },
                { assert workflow.out.target_gnomad_tbi },
                { assert workflow.out.target_mills_vcf },
                { assert workflow.out.target_mills_tbi },
                { assert workflow.out.target_1000g_vcf },
                { assert workflow.out.target_1000g_tbi },
                { assert workflow.out.target_dbsnp_vcf },
                { assert workflow.out.target_dbsnp_tbi },
                { assert workflow.out.clinvar_benign_vcf },
                { assert workflow.out.clinvar_pathogenic_vcf },
                { assert workflow.out.clinvar_selected_vcf },
                { assert workflow.out.versions }
           )
        }
    }
}
