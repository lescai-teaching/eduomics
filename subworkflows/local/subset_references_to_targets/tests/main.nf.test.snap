{
    "human - fasta/BED/VCF - subset by chr": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "10": [
                    
                ],
                "11": [
                    
                ],
                "12": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    
                ],
                "9": [
                    
                ],
                "benign_vcf": [
                    
                ],
                "capture_bed_gz": [
                    
                ],
                "capture_bed_index": [
                    
                ],
                "fa": [
                    
                ],
                "fai": [
                    
                ],
                "pathogenic_vcf": [
                    
                ],
                "selected_tbi": [
                    
                ],
                "selected_vcf": [
                    
                ],
                "sizes": [
                    
                ],
                "target_bed": [
                    
                ],
                "target_bed_pad50": [
                    
                ],
                "target_bed_pad500": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-27T14:28:10.368169865"
    }
}